lib/registries/addon/services/share-search.ts

```diff
@@ -0,0 +1,209 @@
+/* tslint:disable:max-classes-per-file */
+import { Map } from 'immutable';
+import config from 'registries/config/environment';
+import Search, { SearchFilter, SearchModifier, SearchOptions, SearchResults } from 'registries/services/search';
+
+abstract class ShareFilter extends SearchFilter {
+    protected ensureFilterKey(query: any): any {
+        const q = query;
+        q.query = query.query || {};
+        q.query.bool = query.query.bool || {};
+        q.query.bool.filter = query.query.bool.filter || [];
+        return q.query.bool.filter;
+    }
+}
+
+export class ShareTermsFilter extends ShareFilter {
+    static type = Symbol('terms-or');
+
+    constructor(key: string, value: string, display: string) {
+        super({
+            type: ShareTermsFilter.type,
+            display,
+            key,
+            value,
+        });
+    }
+
+    apply(query: any): any {
+        let filter = this.ensureFilterKey(query).find(
+            (el: any) => Object.keys(el.terms || {})[0] === this.internal.get('key'),
+        );
+
+        if (!filter) {
+            filter = { terms: { [this.key]: [] } };
+            query.query.bool.filter.push(filter);
+        }
+
+        filter.terms[this.key].push(this.value);
+
+        return query;
+    }
+}
+
+export class ShareTermsAggregation implements SearchModifier {
+    /* eslint-disable-next-line no-useless-constructor, no-empty-function  */
+    constructor(private name: string, private field: string, private size: number = 200) { }
+
+    apply(query: any): any {
+        const q = query;
+        q.aggregations = q.aggregations || {};
+        q.aggregations[this.name] = {
+            terms: {
+                field: this.field,
+                size: this.size,
+            },
+        };
+        return q;
+    }
+
+    equals(other: any) {
+        return Map(Object.entries(this)).equals(other);
+    }
+
+    hashCode() {
+        return Map(Object.entries(this)).hashCode();
```

Comment:
Why not make a map in the constructor and save it? Or, this seems like a good use for [`Record`](https://facebook.github.io/immutable-js/docs/#/Record).

Reply:
`Record` is actually super weird. It doesn't seem to inherit from `Collection` so it's `hashCode` and `equals` methods don't work and then it just breaks all assumptions one would have about `Set`, `Map`, etc. 

As for the former, probably just laziness? Can/will change.